src/Development/IDE/Core/IdeConfiguration.hs

```diff
@@ -39,24 +44,35 @@ getIdeConfiguration =
 
 parseConfiguration :: InitializeParams -> IdeConfiguration
 parseConfiguration InitializeParams {..} =
-  IdeConfiguration { .. }
+  IdeConfiguration {..}
  where
   workspaceFolders =
     foldMap (singleton . toNormalizedUri) _rootUri
       <> (foldMap . foldMap)
            (singleton . parseWorkspaceFolder)
            _workspaceFolders
+  clientSettings = _initializationOptions
 
 parseWorkspaceFolder :: WorkspaceFolder -> NormalizedUri
 parseWorkspaceFolder =
   toNormalizedUri . Uri . (_uri :: WorkspaceFolder -> Text)
 
 modifyWorkspaceFolders
   :: IdeState -> (HashSet NormalizedUri -> HashSet NormalizedUri) -> IO ()
-modifyWorkspaceFolders ide f = do
+modifyWorkspaceFolders ide f = modifyIdeConfiguration ide f'
+  where f' (IdeConfiguration ws initOpts) = IdeConfiguration (f ws) initOpts
+
+modifyClientSettings
+  :: IdeState -> (Maybe Value -> Maybe Value) -> IO ()
+modifyClientSettings ide f = modifyIdeConfiguration ide f'
+  where f' (IdeConfiguration ws clientSettings) = IdeConfiguration ws (f clientSettings)
+
+modifyIdeConfiguration
+  :: IdeState -> (IdeConfiguration -> IdeConfiguration) -> IO ()
+modifyIdeConfiguration ide f = do
   IdeConfigurationVar var <- getIdeGlobalState ide
-  IdeConfiguration    ws  <- readVar var
-  writeVar var (IdeConfiguration (f ws))
+  ideConfig <- readVar var
+  writeVar var (f ideConfig)
```

Comment:
Let's take the chance to fix the potential race condition in this code by using `modifyVar`